### ISSUE TITLE
thoughtIndexUpdates after cursor movement

### ISSUE THREAD

**Issue by raineorshine** [cybersemics/em#1178]
Why is `updateThoughts` called with a non-empty `thoughtIndexUpdates` on cursor move?

## Steps to Reproduce

```
- a
- b
```

Add `console.log(thoughtIndexUpdates)` to `updateThoughts` reducer to see when it is called.

## Current Behavior

When moving the cursor up and down between `a` and `b`, each move calls `updateThoughts` with a `thoughtIndexUpdates` entry.

## Expected Behavior

The `thoughtIndex` should not be updated from a cursor move.

**Comment by onurpolattimur** [comment]
Here is what I have found so far. 

- Stack trace from pullQueue middleware:
pullQueue → updatePullQueue → flushPullQueue → pull → updateThoughts
- There is an if condition about visible contexts in updatePullQueue. If the last visible contexts equal to current visible contexts, it returns and doesn't call the flushPullQueue. But in this case, it doesn't return because the last visible contexts don't equal current visible contexts.
https://github.com/cybersemics/em/blob/62e29511028cf880e5732568f69d6177344b21f4/src/redux-middleware/pullQueue.ts#L145
- I am not sure that getVisibleContext function returns the context that is really visible. I will try to explain with an example.

```
- a
- b
```

When I move cursor  from `a` to `b`, getVisibleContext returns `["__ROOT__"]` and `["b"]`. Is this the expected result? 
Because `a` is also still visible, shouldn't it return `["__ROOT__"]`, `["b"] `and `["a"]` ?

**Comment by raineorshine** [comment]
> When I move cursor from `a` to `b`, getVisibleContext returns `["__ROOT__"]` and `["b"]`. Is this the expected result?
> Because `a` is also still visible, shouldn't it return `["__ROOT__"]`, `["b"] `and `["a"]` ?

It is a bit confusing: `getVisibleContexts` returns the visible *contexts*, not the visible thoughts. Thus, if `['__ROOT__']` is returned by `getVisibleContexts`, that means that all its children are visible. It doesn't add each of the children individually.

- When the cursor is on `a` in this example the visible contexts are `['__ROOT__']` and `['a']` (because if `a` had any children they would be visible while b's would not be)
- When the cursor is on `b` the visible contexts are `['__ROOT__']` and `['b']`

`getVisibleContexts` is used to determine if the `pullQueue` should be flushed (i.e. all visible thoughts loaded from the database). If it is flushing unnecessarily, it may be a performance issue but in itself it doesn't update thoughts.

The problem is likely in `pull`, `updateThoughts`, or `reconcile` (apologies for the state of that code 😅). It may even be exactly what is mentioned in this comment:

https://github.com/cybersemics/em/blob/db3afa73a3fc8a28bab5b478dcc6531a8d1eb70b/src/action-creators/pull.ts#L42-L48

Aside from pursuing those functions, there is an question which is important to assess the impact of this idiosyncrasy: Is the update redundant (performance problem)? Or is it changing the state (bug)?

Good luck 😄

**Comment by raineorshine** [comment]
@shresthabijay Could you take over this please?

**Comment by shresthabijay** [comment]
> @shresthabijay Could you take over this please?

Sure. I will take a look.

**Comment by shresthabijay** [comment]
https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L37-L55

For every `thoughts` in `thoughtsLocalIterable`, `updateThoughts` is being called. I believe the reason for calling `updateThoughts` each time is to make sure that the application state is synced with local pull updates even if the remote pull is not available.

My question is why can't we `updateThoughts` with merged local thoughts updates only when the remote is not available for sync. If the remote pull is available we are reconciling the local and remote updates anyways.

**Comment by shresthabijay** [comment]
https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L66-L96

Why are we reconciling here thoughts updates that are available both locally and remotely here. Is it all handled here by the reconciliation

https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L104-L07

If it does the same thing here, why redundant reconciliation in the above for loop?

**Comment by shresthabijay** [comment]
> `getVisibleContexts` is used to determine if the `pullQueue` should be flushed (i.e. all visible thoughts loaded from the database). If it is flushing unnecessarily, it may be a performance issue but in itself it doesn't update thoughts.

I think it should be addressed too. We may be pulling data unnecessarily with every change of visible contexts. We should only pull the data that has not been pulled already. Unless we are trying to get new updates for already existing data from this method.
Are there plans to sync multiple devices using firebase data updates?

**Comment by raineorshine** [comment]
> My question is why can't we `updateThoughts` with merged local thoughts updates only when the remote is not available for sync. If the remote pull is available we are reconciling the local and remote updates anyways.

Presumably local thoughts load much faster than remote thoughts. If local thoughts take 200ms to load, and remote thoughts take 500ms to load, that will be a noticeable different to the user. We want the local thoughts to load as soon as they are ready.

That is even assuming a stable connection. If the user's internet connection is unstable, they may have local thoughts that are not loaded at all as it hangs and waits for remote thoughts to load.

(FYI Batch loading from IndexedDB is way slower than I thought, but we haven't tried to optimize it yet, and remote will always be slower than local in principle.)

> Here we are reconciling thoughts updates that are available both locally and remotely. Isn't it all handled here by this reconciliation?
> 
> https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L104-L107
> 
> If it does the same thing here, why redundant reconciliation in the above for loop?

It's kind of explained in the comment, but let me try to explain better.

```
    // the reconcile dispatched above is based on remote keys only
    // thoughts that exist locally and not remotely will be missed
    // sync all thoughts here to ensure none are missed
    // TODO: Only reconcile local-only thoughts
```

I think this last `reconcile` syncs the local thoughts back to the remote. The [first updateThoughts](https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L44) only updates state, the [reconcile in the remote loop](https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L87) only reconciles parents/lexemes that exist in both local and remote. 

The [final reconcile](https://github.com/cybersemics/em/blob/a2b8ad569074b73023cabca69144b05029c76151/src/action-creators/pull.ts#L104) should sync contexts/lexemes that are local but not in the remote. Clearly it is not so precise, hence the TODO.

No wonder there are redundant calls to `updateThoughts`... I lost steam at the end of my `pull` implementation and didn't filter down the correct updates for a precise reconcile.

> > `getVisibleContexts` is used to determine if the `pullQueue` should be flushed (i.e. all visible thoughts loaded from the database). If it is flushing unnecessarily, it may be a performance issue but in itself it doesn't update thoughts.
> 
> I think it should be addressed too. We may be pulling data unnecessarily with every change of visible contexts. We should only pull the data that has not been pulled already. Unless we are trying to get new updates for already existing data from this method.
> Are there plans to sync multiple devices using firebase data updates?

Yes, we have to assume that the user is connected with multiple devices and pull updates before we update the remote.

That said, @anmolarora1 is working on #932 so that we can subscribe to Firebase changes and not rely on a pre-push pull. It is a Data Integrity issue because there is a bug that stems from a race condition between all the pushes and pulls that are generated.

**Comment by raineorshine** [comment]
@shresthabijay Thoughts?

Also, I’m still looking for a concise answer this question when you have time. Thanks!

> Aside from pursuing those functions, there is a question which is important to assess the impact of this idiosyncrasy: Is the update redundant (performance problem)? Or is it changing the state (bug)?

Let’s confirm exactly what changes to the state it is making if any. If it is incorrectly updates the time stamp, for example, that may cause problems in reconcile.

**Comment by shresthabijay** [comment]
@raineorshine I have some thoughts about the actual cause of this issue. Why is the cursor movement triggering pull in the first place ?

 `pullQueue` adds the context to pull whenever `expandThoughts` changes. However `expandThoughts` changes on cursor movement. I think change in the `expandThoughts` should not be the entire reason for triggering `pull`. Current implementation feels like we are constantly trying to fetch updated data from remote and reconcile with local on every cursor movement due to lack of remote subscribtion and sync.

If these things are in place then  the `pullQueue` should only be responsible for pulling pending contexts.
- Subscribing remote changes to sync changes whenever online.
- Mechanism for offline device to sync local data with the remote when a device comes back online.

The main question is when should `pullQueue` trigger `pull` ? According to my understanding these are the conditions:
- When there are pending thoughts among the current visible context and their children. (Dependent on expandThoughts but needs more calculation to actually conclude if pull is necessary)
- When offline and there is an active search and it changes.

**Comment by shresthabijay** [comment]
> Let’s confirm exactly what changes to the state it is making if any. If it is incorrectly updating the time stamp, for example, that may cause problems in reconcile.

I used redux dev tools to analyze the diff. I logged in and after all the reconciliation,  I activated `cursorDown` to trigger the redundant pull.  With all the redundant updates, the only thing that changes in the state is the `thoughtCache`. The `expandThoughts` and `thoughts` remain the same even though they are recalculated. I am not sure about the impact of the change in `thoughtCache` due to these redundant updates. Other than that I think it is a performance problem.

**Comment by raineorshine** [comment]
> `pullQueue` adds the context to pull whenever `expandThoughts` changes. However `expandThoughts` changes on cursor movement. I think change in the `expandThoughts` should not be the entire reason for triggering `pull`. Current implementation feels like we are constantly trying to fetch updated data from remote and reconcile with local on every cursor movement due to lack of remote subscribtion and sync.

Yes, that's correct. Once we have db subscriptions working, we won't have to pull every time the cursor changes (though we will have to update the subscriptions).

> * Mechanism for offline device to sync local data with the remote when a device comes back online.

The good news: Firebase queues changes automatically when offline and resubmits them when the user comes back online. The bad news: If the remote has changed, the queued updates could be out of sync with the remote state and theoretically create data integrity issues on the remote.

> The main question is when should `pullQueue` trigger `pull` ? According to my understanding these are the conditions:
> 
> * When there are pending thoughts among the current visible context and their children. (Dependent on expandThoughts but needs more calculation to actually conclude if pull is necessary)
> * When offline and there is an active search and it changes.

The `pullQueue` was originally designed to `pull` thoughts whenever the visible thoughts change. With subscriptions, it will only have to pull 1) on load and 2) on navigating to load pending thoughts (normal view, context view, and search).

Should we wait until #932 is merged before addressing this?

How do we deal with the potential data integrity issues when coming back online and pushing changes to the remote? Does Firebase have a way to handle this? We can't be the only ones syncing multiple devices over Firebase. Maybe they have validation rules that prevents data integrity issues, and expect the client to roll back those changes. That would be very difficult for us since they would already be in `IndexedDB`.

I'm a bit stumped by this whole push/pull/reconcile situation 🤔

**Comment by shresthabijay** [comment]
> Should we wait until #932 is merged before addressing this?

I agree on this. After this PR, `pull` is not required after every cursor movement.

**Comment by shresthabijay** [comment]
> How do we deal with the potential data integrity issues when coming back online and pushing changes to the remote? Does Firebase have a way to handle this? We can't be the only ones syncing multiple devices over Firebase. Maybe they have validation rules that prevents data integrity issues, and expect the client to roll back those changes. That would be very difficult for us since they would already be in `IndexedDB`.

I don't know if the Firebase handles this. I would get back to you after I do some research on this.

**Comment by raineorshine** [comment]
`pull` now only pulls pending thoughts (some exceptions). Fixed in #1133 and a4bfc8e63b.

I'll close this issue and we can open a new issue for the syncing on reconnection problem when we have a reproducible case.
